src/crypto/crypto_kmac.cc

```diff
@@ -0,0 +1,217 @@
+#include "crypto/crypto_kmac.h"
+#include "async_wrap-inl.h"
+#include "node_internals.h"
+#include "threadpoolwork-inl.h"
+
+#if OPENSSL_VERSION_MAJOR >= 3
+#include <openssl/core_names.h>
+#include <openssl/params.h>
+#include "crypto/crypto_keys.h"
+#include "crypto/crypto_sig.h"
+#include "ncrypto.h"
+
+namespace node::crypto {
+
+using ncrypto::EVPMacCtxPointer;
+using ncrypto::EVPMacPointer;
+using node::Utf8Value;
+using v8::Boolean;
+using v8::FunctionCallbackInfo;
+using v8::JustVoid;
+using v8::Local;
+using v8::Maybe;
+using v8::MaybeLocal;
+using v8::Nothing;
+using v8::Object;
+using v8::Uint32;
+using v8::Value;
+
+KmacConfig::KmacConfig(KmacConfig&& other) noexcept
+    : job_mode(other.job_mode),
+      mode(other.mode),
+      key(std::move(other.key)),
+      data(std::move(other.data)),
+      signature(std::move(other.signature)),
+      customization(std::move(other.customization)),
+      algorithm_name(std::move(other.algorithm_name)),
+      length(other.length) {}
+
+KmacConfig& KmacConfig::operator=(KmacConfig&& other) noexcept {
+  if (&other == this) return *this;
+  this->~KmacConfig();
+  return *new (this) KmacConfig(std::move(other));
+}
+
+void KmacConfig::MemoryInfo(MemoryTracker* tracker) const {
+  tracker->TrackField("key", key);
+  // If the job is sync, then the KmacConfig does not own the data.
+  if (job_mode == kCryptoJobAsync) {
+    tracker->TrackFieldWithSize("data", data.size());
+    tracker->TrackFieldWithSize("signature", signature.size());
+    tracker->TrackFieldWithSize("customization", customization.size());
+  }
+}
+
+Maybe<void> KmacTraits::AdditionalConfig(
+    CryptoJobMode mode,
+    const FunctionCallbackInfo<Value>& args,
+    unsigned int offset,
+    KmacConfig* params) {
+  Environment* env = Environment::GetCurrent(args);
+
+  params->job_mode = mode;
+
+  CHECK(args[offset]->IsUint32());  // SignConfiguration::Mode
+  params->mode =
+      static_cast<SignConfiguration::Mode>(args[offset].As<Uint32>()->Value());
+
+  CHECK(args[offset + 1]->IsObject());  // Key
+  KeyObjectHandle* key;
+  ASSIGN_OR_RETURN_UNWRAP(&key, args[offset + 1], Nothing<void>());
+  params->key = key->Data().addRef();
+
+  CHECK(args[offset + 2]->IsString());  // Algorithm name
+  Utf8Value algorithm_name(env->isolate(), args[offset + 2]);
+  params->algorithm_name = std::string(*algorithm_name);
```

Comment:
```suggestion
  params->algorithm_name = algorithm_name.ToString();
```

(avoids string scan for length + will never handle nul bytes incorrectly)